### DISCUSSION TITLE
ESP32 + ILI9488 working but not quite right

### DISCUSSION THREAD

**Discussion by JayNaire** [Bodmer/TFT_eSPI#2529] [Q&A - Touch]
linux->vscode-platformio->esp32doit-devkit-v1 espressif32 

I had an ESP32 + ILI9341 working perfectly.  uint16_t calData[5] = { 393, 3164, 265, 3356, 2 }
 Using _examples/Generic/Touch_calibrate/Touch_calibrate.ino_ the dots are drawn on the screen precisely at the tip of the stylus.

I substituted an ILI9488 screen for the ILI9341. ( uint16_t calData[5] = { 370, 2747, 327, 3386, 6 } )
Everything works except that the dots don't correspond to the stylus. The further to the left of the screen the stylus is, the further to the left of the stylus the dot is drawn.  Max. distance is ~10mm resulting in a "dead" 10mm margin on the left of the screen. A similar thing happens vertically except the top margin is about 5mm.

Coordinates reported to Serial bear this behaviour out eg: x,y = 0,269 occurs when the stylus is 10mm from the left screen edge.

If you touch the stylus inside the dead margin random dots start appearing roughly at the right y coordinate but incorrect x.

I tried another ILI9488 screen with the same results.

Does this behaviour ring any bells please?

**Comment by Bodmer** [comment]
Have you run a touch calibration sketch?

**Reply by JayNaire** [reply]
Yes. Figures are as given in the post.

**Comment by Bodmer** [comment]
Is SDO disconnected as noted here:
https://github.com/Bodmer/TFT_eSPI/discussions/898

**Reply by JayNaire** [reply]
Yes. (It was originally connected and touch didn't work at all.)
Additional very tentative info:
1. When testing, the max. discrepancy between drawn dot and stylus seems to be affected by where in the corner boxes I touched the stylus when calibrating. I managed to reduce this discrepancy to about 2mm as opposed to 10mm by touching a little way in from the tip of the calibration arrow. I have not tried to repeat this. This may be coincidence and a total red herring.

2. You can roughly predict where a dot will be drawn. Imagine radial lines extending from the screen centre. Touch along a radial. The dot will appear along that radial some distance toward the screen border. That is, x and y have scaled by an amount. That amount depends on how far the stylus hits from the centre.

**Comment by Bodmer** [comment]
Run this sketch and look at the output in a serial monitor window when corners are touched.
https://github.com/Bodmer/TFT_eSPI/tree/master/examples/Test%20and%20diagnostics/Test_Touch_Controller
 This will report the raw ADC values. What are the Z pressure and x,y values in the corners and in the centre?

**Reply by JayNaire** [reply]
Many thanks. Here's what I found from testing two ILI9488 screens:

**Raw ADC output screen board A**
| ILI9488*      |  X    |    Y    |    Z   |
|--------------|-------|---------|--------|
|Top left:     | 3455  |   3746  |   885  |
|Top right:    | 374   |   3621  |   225  |
|Bottom left:  | 3339  |   272   |   1627 |
|Bottom right: | 396   |   289   |   1265 |
|~Centre:       | 1940  |   2051  |   2092 |

**Raw ADC output screen board B**
| ILI9488      |  X    |    Y    |    Z   |
|--------------|-------|---------|--------|
|Top left:     | 3225  |   3647  |   709  |
|Top right:    | 384   |   3598  |   134  |
|Bottom left:  | 3124  |   329   |   1398 |
|Bottom right: | 432   |   259   |   1549 |
|~Centre:       | 1989  |   2082  |   2180 |

*Notes:
1.  ILI9488 is physically mounted 180deg wrt connector cf. ILI9341
2. tft.setRotation(2) corrects the orientation but not the discrepancy between stylus and dot position
